Python/lock.c

```diff
@@ -353,3 +353,107 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
         v = _Py_atomic_load_uint8(&flag->v);
     }
 }
+
+#define _Py_WRITE_LOCKED 1
+#define _PyRWMutex_READER_SHIFT 2
+#define _Py_RWMUTEX_MAX_READERS (UINTPTR_MAX >> _PyRWMutex_READER_SHIFT)
+
+static uintptr_t
+rwmutex_set_parked_and_wait(_PyRWMutex *rwmutex, uintptr_t bits)
+{
+    // Set _Py_HAS_PARKED and wait until we are woken up.
+    if ((bits & _Py_HAS_PARKED) == 0) {
+        uintptr_t newval = bits | _Py_HAS_PARKED;
+        if (!_Py_atomic_compare_exchange_uintptr(&rwmutex->bits,
+                                                 &bits, newval)) {
+            return bits;
+        }
+        bits = newval;
+    }
+
+    _PyParkingLot_Park(&rwmutex->bits, &bits, sizeof(bits), -1, NULL, 1);
+    return _Py_atomic_load_uintptr_relaxed(&rwmutex->bits);
+}
+
+// The number of readers holding the lock
+static uintptr_t
+rwmutex_reader_count(uintptr_t bits)
+{
+    return bits >> _PyRWMutex_READER_SHIFT;
+}
+
+void
+_PyRWMutex_RLock(_PyRWMutex *rwmutex)
+{
+    uintptr_t bits = _Py_atomic_load_uintptr_relaxed(&rwmutex->bits);
+    for (;;) {
+        if ((bits & _Py_WRITE_LOCKED)) {
+            // The lock is write-locked.
+            bits = rwmutex_set_parked_and_wait(rwmutex, bits);
+            continue;
+        }
+        else if ((bits & _Py_HAS_PARKED)) {
+            // There is at least one waiting writer. We can't grab the lock
+            // because we don't want to starve the writer. Instead, we park
+            // ourselves and wait for the writer to eventually wake us up.
+            bits = rwmutex_set_parked_and_wait(rwmutex, bits);
+            continue;
+        }
```

Comment:
This branch is called when readers hold the lock and there is a waiting writer. Although `_Py_HAS_PARKED` in general can indicate waiting readers or writers, if the lock is read-locked it must imply a waiting writer because otherwise readers don't block other readers.

Reply:
Ah, I get it now.  Thanks for explaining.

Reply:
There's actually one other case that this branch handles that I forgot about: There's a transient state after the last reader unlocks the mutex but before a writer grabs it in which `_Py_HAS_PARKED` is set, but the lock is neither read-locked nor write-locked. It's a transient state because there's guaranteed to be at least one previously waiting writer actively trying to write-lock the mutex.

The behavior is the same in this state -- we still want readers to wait for the writer -- but it means the assertion I added did not hold.